### ISSUE TITLE
Hi, which file should be run to decode wenet ?

### ISSUE THREAD

**Issue by jinfagang** [k2-fsa/sherpa#139]
https://github.com/k2-fsa/sherpa/tree/master/sherpa/bin

all of them using bpe model but I only got token.txt

**Comment by csukuangfj** [comment]
Please read the documentation or the code for usage.

All of them also support --token-filename.

**Comment by jinfagang** [comment]
@csukuangfj Which file should using? doc is very confusing. Also the bin doesn't seems have a token-filename

**Comment by csukuangfj** [comment]
What do you want to do?

**Comment by jinfagang** [comment]
@csukuangfj  this part I get confusing:

```
 params.blank_id = sp.piece_to_id("<blk>")
    params.vocab_size = sp.get_piece_size()
```

HOw can I get these using token_table then?

**Comment by csukuangfj** [comment]
> Hi, which file should be run to decode wenet

The wenetspeech model is for Chinese.

You can find doc for how to run Chinese models at https://k2-fsa.github.io/sherpa/python/streaming_asr/index.html

The above page describes how to run sherpa for each kind of supported model.

I assume that you know what type of model you are using, e.g., lstm-transducer, conformer-transducer, emformer-transducer, etc.

---

> Which file should using? doc is very confusing

Sorry about that. Could you be more specific?

**Comment by csukuangfj** [comment]
> @csukuangfj this part I get confusing:
> 
> ```
>  params.blank_id = sp.piece_to_id("<blk>")
>     params.vocab_size = sp.get_piece_size()
> ```
> 
> HOw can I get these using token_table then?

There are no such pieces of code in sherpa.

---

I assume you are referring to icefall. If so, this kind of code is specific for BPE-based models.

Every model from icefall has its own directory and you can find how to use that model by looking at its `pretrained.py`.

**Comment by jinfagang** [comment]
Yes, this code comes from icefall. I get very miss around by swithcing your guys tools.... I just take a workable version nomatter where it comes from. 

At least, it make me understand each one of step by step.

Now, things get confused, Does all  Chinese not using BPE models?

If no BPE, how sohuld I get blank_id? 

In icefall, it comes from:

```
 params.blank_id = sp.piece_to_id("<blk>")
    params.vocab_size = sp.get_piece_size()
```

```
decoder = Decoder(
        vocab_size=params.vocab_size,
        decoder_dim=params.decoder_dim,
        blank_id=params.blank_id,
        context_size=params.context_size,
    )
```

**Comment by csukuangfj** [comment]
> Now, things get confused, Does all Chinese not using BPE models?

Yes, no Chinese model uses BPE models. I suggest that you search for what BPE is if you are interested.

---

> In icefall, it comes from:
>
> params.blank_id = sp.piece_to_id("<blk>")
 >   params.vocab_size = sp.get_piece_size()

I assume you are looking at the code using BPE models in icefall, e.g., librispeech.

I suggest that you look at the code using tokens.txt instead if you want to use models for Chinese. For instance, 
https://github.com/k2-fsa/icefall/blob/master/egs/wenetspeech/ASR/pruned_transducer_stateless2/train.py

https://github.com/k2-fsa/icefall/blob/9ae2f3a3c5a3c2336ca236c984843c0e133ee307/egs/wenetspeech/ASR/pruned_transducer_stateless2/train.py#L830

```python3
    params.blank_id = lexicon.token_table["<blk>"]
    params.vocab_size = max(lexicon.tokens) + 1
```

**Comment by jinfagang** [comment]
@csukuangfj I am confused...... I don't need Lexicon.... I just want token.....

HOw to use this: `token_table = k2.SymbolTable.from_file(args.bpe_model)`........

**Comment by csukuangfj** [comment]
All the information you need can be found from `token.txt`, i.e., `token_table`.

Please refer to how `lexicon.tokens` is implemented
https://github.com/k2-fsa/icefall/blob/master/icefall/lexicon.py#L190

You can find the help info about `token_table` at
https://k2-fsa.github.io/k2/python_api/api.html#symboltable

----

> I get very miss around by swithcing your guys tools.

We have documentation for each project/tool:

- https://k2-fsa.github.io/k2
- https://k2-fsa.github.io/icefall
- https://k2-fsa.github.io/sherpa
- https://lhotse.readthedocs.io/

If you still don't know how to do it after reading the doc, please describe first what you have done and what is not clear from the doc and we will update the doc.

**Comment by jinfagang** [comment]
@csukuangfj  
```
params.blank_id = sp['<blk>']
        params.vocab_size = max(sp) + 1
        logger.info(f'vocab size: {params.vocab_size}')

2/symbol_table.py", line 199, in get
    return self._id2sym[k]
KeyError: 5539
```

**Comment by jinfagang** [comment]
@csukuangfj  I directly using token txt SymbolTable, but get above error.

**Comment by csukuangfj** [comment]
Where did you download the `tokens.txt`? And which model are you using?
`tokens.txt` for models with WenetSpeech cannot be used for models using BPEs and vice versa.

**Comment by csukuangfj** [comment]
I strongly recommend that you first follow https://k2-fsa.github.io/sherpa
to run the model you want to use with sherpa. 

After you get it working, you can port the code to your project and run the model with your project.

**Comment by jinfagang** [comment]
@csukuangfj  I error comes from  `        params.vocab_size = max(sp) + 1
` not run into model yet....

I think the only difference is the vocabsize ??

**Comment by jinfagang** [comment]
@csukuangfj  I can not get your suggest code work:

```
token_table = k2.SymbolTable.from_file(args.bpe_model)
        vc  = max(token_table) + 1
        print(vc)
```

Please don't make me miss around... It should be simple enough question...

**Comment by csukuangfj** [comment]
> token_table = k2.SymbolTable.from_file(args.bpe_model)

This is not right.
please see https://github.com/k2-fsa/sherpa/blob/c235538f36bcf59023caecfe594fa9ee2c6177cd/sherpa/bin/lstm_transducer_stateless/streaming_server.py#L269

---

> Please don't make me miss around... It should be simple enough question...

Since you are porting code to your own project, I assume that you know  at least how does the code work.
If not, I recommend you spend some time reading the doc of sherpa and icefall first.

**Comment by jinfagang** [comment]
<img width="535" alt="image" src="https://user-images.githubusercontent.com/21303438/192138066-a9f563cd-49d7-4616-81ee-f78b8d13875d.png">

how does model can get blank_id and vocabsize 
/?

**Comment by jinfagang** [comment]
```
self.model = RnntLstmModel(
            nn_encoder_filename,
            nn_decoder_filename,
            nn_joiner_filename,
            device=device,
        )

```

can't even nkonw where does it send token file

**Comment by csukuangfj** [comment]
> how does model can get blank_id and vocabsize

First, you don't need to pass blank_id and vocab size to models in icefall.  They are attributes of the model when they are exported from pytorch using torchscript.

**Comment by csukuangfj** [comment]
If you want to get blank_id and vocab size from the token table, please see my comments in the above.
https://github.com/k2-fsa/sherpa/issues/139#issuecomment-1257154737

https://github.com/k2-fsa/sherpa/issues/139#issuecomment-1257152481
